fearless/ApplicationLayer/Services/WalletBalanceFetcher.swift

```diff
@@ -0,0 +1,405 @@
+
+    private lazy var accountInfosAdapters: [String: AccountInfoSubscriptionAdapter] = [:]
+    private lazy var accountInfos: [ChainAssetKey: AccountInfo] = [:]
+    private lazy var chainAssets: [ChainAsset] = []
+    private lazy var metaAccounts: [MetaAccountModel] = []
+    private lazy var prices: [PriceData] = []
+
+    // MARK: - Constructor
+
+    init(
+        metaAccountRepository: AnyDataProviderRepository<MetaAccountModel>,
+        priceLocalSubscriptionFactory: PriceProviderFactoryProtocol,
+        chainRepository: AnyDataProviderRepository<ChainModel>,
+        operationQueue: OperationQueue,
+        logger: Logger
+    ) {
+        self.priceLocalSubscriptionFactory = priceLocalSubscriptionFactory
+        self.metaAccountRepository = metaAccountRepository
+        self.chainRepository = chainRepository
+        self.operationQueue = operationQueue
+        self.logger = logger
+    }
+
+    // MARK: - Public methods
+
+    func fetchBalanceFor(
+        walletId: String?,
+        deliverOn queue: DispatchQueue?,
+        completion: @escaping WalletBalanceResultCompletion
+    ) {
+        self.completion = completion
+        deliverQueue = queue
+
+        if let identifier = walletId {
+            fetchMetaAccount(by: identifier)
+        } else {
+            fetchAllMetaAccounts()
+        }
+    }
+
+    // MARK: - Private methods
+
+    private func buildBalance() {
+        guard
+            accountInfos.isNotEmpty,
+            prices.isNotEmpty
+        else {
+            return
+        }
+
+        let builderBlock = BlockOperation {
+            let walletBalances = self.walletBalanceBuilder.buildBalance(
+                accountInfos: self.accountInfos,
+                metaAccounts: self.metaAccounts,
+                chainAssets: self.chainAssets,
+                prices: self.prices
+            )
+
+            self.deliverOnCompletion(.success(walletBalances))
+        }
+
+        operationQueue.addOperation(builderBlock)
+    }
+
+    private func handle(_ wallets: [MetaAccountModel], _ chains: [ChainModel]) {
+        let chainsAssets = chains.map(\.chainAssets).reduce([], +)
+        chainAssets = chainsAssets
+        metaAccounts = wallets
+
+        subscribeToAccountInfo(for: wallets, chainsAssets)
+        subscribeToPrices(for: chains)
+    }
+
+    private func fetchMetaAccount(by identifire: String) {
+        let metaAccountOperation = metaAccountRepository.fetchOperation(
+            by: identifire,
+            options: RepositoryFetchOptions()
+        )
+        let chainsOperation = fetchChainsOperation()
+
+        metaAccountOperation.addDependency(chainsOperation)
+
+        metaAccountOperation.completionBlock = { [weak self] in
+            guard
+                let metaAccountResult = metaAccountOperation.result,
+                let chainsResult = chainsOperation.result
+            else {
+                self?.deliverOnCompletion(.failure(WalletBalanceError.noData))
```

Comment:
it will be better to specify missing data. e.g.
.accountMissing
.chainsMissing